### ISSUE TITLE
Mist v 1.2 - Building Catalina on Big Sur fails

### ISSUE THREAD

**Issue by carlashley** [ninxsoft/Mist#16]
When building a Catalina app on an M1 (Big Sur 11.4) with `mist`, it fails. This happened even when specifying the build (macOS 10.15.7 Build 19H15).

Build unspecified:
```
sudo ./mist -d -n Catalina -o /tmp/ -a 
[INFO] 2021-05-31 00:00:42 +0000 - Checking for macOS with name 'Catalina', version 'latest' and build 'latest'...
[INFO] 2021-05-31 00:00:42 +0000 - Downloading catalog https://swscan.apple.com/content/catalogs/others/index-10.16-10.15-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog...
[INFO] 2021-05-31 00:00:43 +0000 - Downloading catalog https://swscan.apple.com/content/catalogs/others/index-10.15-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog...
[INFO] 2021-05-31 00:00:44 +0000 - Downloading catalog https://swscan.apple.com/content/catalogs/others/index-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog...
[INFO] 2021-05-31 00:00:44 +0000 - Downloading catalog https://swscan.apple.com/content/catalogs/others/index-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog...
[INFO] 2021-05-31 00:00:45 +0000 - Found macOS Catalina 10.15.7 (19H15)...
[INFO] 2021-05-31 00:00:45 +0000 - Creating new temporary directory '/private/var/tmp/001-68446'...
[INFO] 2021-05-31 00:00:45 +0000 - Created new temporary directory '/private/var/tmp/001-68446'
[INFO] 2021-05-31 00:00:45 +0000 - [01 / 12] Downloading 001-68446.English.dist...[08.97 KB of 08.97 KB - 100.0%]
[INFO] 2021-05-31 00:00:45 +0000 - [02 / 12] Downloading AppleDiagnostics.chunklist...[28.00 B of 28.00 B - 100.0%]
[INFO] 2021-05-31 00:00:45 +0000 - [03 / 12] Downloading AppleDiagnostics.dmg...[03.15 MB of 03.15 MB - 100.0%]
[INFO] 2021-05-31 00:00:45 +0000 - [04 / 12] Downloading BaseSystem.chunklist...[02.02 KB of 02.02 KB - 100.0%]
[INFO] 2021-05-31 00:00:45 +0000 - [05 / 12] Downloading BaseSystem.dmg...[98.63 MB of 98.63 MB - 100.0%]
[INFO] 2021-05-31 00:00:50 +0000 - [06 / 12] Downloading InstallAssistantAuto.pkg...[10.75 MB of 10.75 MB - 100.0%]
[INFO] 2021-05-31 00:00:51 +0000 - [07 / 12] Downloading InstallESDDmg.chunklist...[26.90 KB of 26.90 KB - 100.0%]
[INFO] 2021-05-31 00:00:51 +0000 - [08 / 12] Downloading InstallESDDmg.pkg...[07.74 GB of 07.74 GB - 100.0%]
[INFO] 2021-05-31 00:02:09 +0000 - [09 / 12] Downloading InstallInfo.plist...[01.58 KB of 01.58 KB - 100.0%]
[INFO] 2021-05-31 00:02:09 +0000 - [10 / 12] Downloading MajorOSInfo.pkg...[01.90 MB of 01.90 MB - 100.0%]
[INFO] 2021-05-31 00:02:09 +0000 - [11 / 12] Downloading OSInstall.mpkg...[99.43 KB of 99.43 KB - 100.0%]
[INFO] 2021-05-31 00:02:09 +0000 - [12 / 12] Downloading RecoveryHDMetaDmg.pkg...[00.66 MB of 00.66 MB - 100.0%]
[INFO] 2021-05-31 00:02:24 +0000 - Creating new installer '/Applications/Install macOS Catalina.app'...
installer: Error - ERROR_DC35364117
[ERROR] 2021-05-31 00:02:24 +0000 - Invalid Exit Status Code: '1', Arguments: ["installer", "-pkg", "/private/var/tmp/001-68446/001-68446.English.dist", "-target", "/"]
Error: invalidExitStatus(code: 1, arguments: ["installer", "-pkg", "/private/var/tmp/001-68446/001-68446.English.dist", "-target", "/"])
```

Specifying the build:
```
sudo ./mist -d -n Catalina -b 19H15 -o /tmp/ -a
[INFO] 2021-05-31 00:22:35 +0000 - Checking for macOS with name 'Catalina', version 'latest' and build '19H15'...
[INFO] 2021-05-31 00:22:35 +0000 - Downloading catalog https://swscan.apple.com/content/catalogs/others/index-10.16-10.15-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog...
[INFO] 2021-05-31 00:22:36 +0000 - Downloading catalog https://swscan.apple.com/content/catalogs/others/index-10.15-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog...
[INFO] 2021-05-31 00:22:37 +0000 - Downloading catalog https://swscan.apple.com/content/catalogs/others/index-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog...
[INFO] 2021-05-31 00:22:37 +0000 - Downloading catalog https://swscan.apple.com/content/catalogs/others/index-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog...
[INFO] 2021-05-31 00:22:38 +0000 - Found macOS Catalina 10.15.7 (19H15)...
[INFO] 2021-05-31 00:22:38 +0000 - Deleting old temporary directory '/private/var/tmp/001-68446'...
[INFO] 2021-05-31 00:22:38 +0000 - Deleted old temporary directory '/private/var/tmp/001-68446'
[INFO] 2021-05-31 00:22:38 +0000 - Creating new temporary directory '/private/var/tmp/001-68446'...
[INFO] 2021-05-31 00:22:38 +0000 - Created new temporary directory '/private/var/tmp/001-68446'
[INFO] 2021-05-31 00:22:38 +0000 - [01 / 12] Downloading 001-68446.English.dist...[08.97 KB of 08.97 KB - 100.0%]
[INFO] 2021-05-31 00:22:38 +0000 - [02 / 12] Downloading AppleDiagnostics.chunklist...[28.00 B of 28.00 B - 100.0%]
[INFO] 2021-05-31 00:22:38 +0000 - [03 / 12] Downloading AppleDiagnostics.dmg...[03.15 MB of 03.15 MB - 100.0%]
[INFO] 2021-05-31 00:22:39 +0000 - [04 / 12] Downloading BaseSystem.chunklist...[02.02 KB of 02.02 KB - 100.0%]
[INFO] 2021-05-31 00:22:39 +0000 - [05 / 12] Downloading BaseSystem.dmg...[98.63 MB of 98.63 MB - 100.0%]
[INFO] 2021-05-31 00:22:50 +0000 - [06 / 12] Downloading InstallAssistantAuto.pkg...[10.75 MB of 10.75 MB - 100.0%]
[INFO] 2021-05-31 00:22:51 +0000 - [07 / 12] Downloading InstallESDDmg.chunklist...[26.90 KB of 26.90 KB - 100.0%]
[INFO] 2021-05-31 00:22:51 +0000 - [08 / 12] Downloading InstallESDDmg.pkg...[07.74 GB of 07.74 GB - 100.0%]
[INFO] 2021-05-31 00:25:53 +0000 - [09 / 12] Downloading InstallInfo.plist...[01.58 KB of 01.58 KB - 100.0%]
[INFO] 2021-05-31 00:25:53 +0000 - [10 / 12] Downloading MajorOSInfo.pkg...[01.90 MB of 01.90 MB - 100.0%]
[INFO] 2021-05-31 00:25:53 +0000 - [11 / 12] Downloading OSInstall.mpkg...[99.43 KB of 99.43 KB - 100.0%]
[INFO] 2021-05-31 00:25:53 +0000 - [12 / 12] Downloading RecoveryHDMetaDmg.pkg...[00.66 MB of 00.66 MB - 100.0%]
[INFO] 2021-05-31 00:26:02 +0000 - Creating new installer '/Applications/Install macOS Catalina.app'...
installer: Error - ERROR_DC35364117
[ERROR] 2021-05-31 00:26:02 +0000 - Invalid Exit Status Code: '1', Arguments: ["installer", "-pkg", "/private/var/tmp/001-68446/001-68446.English.dist", "-target", "/"]
Error: invalidExitStatus(code: 1, arguments: ["installer", "-pkg", "/private/var/tmp/001-68446/001-68446.English.dist", "-target", "/"])
```

**Comment by ninxsoft** [comment]
Without any testing whatsoever, my quick assumptions are as follows:

* Certain macOS Installers (ie. 10.15.7 19H4, 10.13.6 17G2208) will only install correctly on certain hardware (ie. 2018 MBP)
* M1 hardware has only ever shipped with macOS Big Sur (or later)
* macOS Catalina (and older) packages most likely have logic within the packages to fail on unsupported / unknown hardware (ie. M1 Macs)

Will need to validate the points above, can definitely improve the smarts around preventing downloads of unsupported builds.

**Comment by carlashley** [comment]
This definitely fails on M1 but works fine on Intel. The `installinstallmacos.py` script by Greg Neagle defaults to creating `dmg` files for all the installers it creates, and I can create a 10.15.7 installer to DMG using `installinstallmacos.py` on an M1 without problems.

**Comment by ninxsoft** [comment]
Having a look at [installinstallmacos.py](https://github.com/munki/macadmin-scripts/blob/main/installinstallmacos.py), and some very quick Googling, I can ascertain the following:

* Setting a `CM_BUILD=CM_BUILD` environment variable should allow **mist** to create specific builds of the macOS Installer (ie. 10.15.7 19H4, 10.13.6 17G2208) on all Mac hardware
* Installing the macOS Installer packages to a sparse disk image should help with **1.** targeting non-booted volumes and **2.** free space requirements
* Will need to validate and work through the post install script bug outlined in [installinstallmacos.py](https://github.com/munki/macadmin-scripts/blob/main/installinstallmacos.py)

**Comment by ninxsoft** [comment]
`CM_BUILD=CM_BUILD` did the trick, fixed in 7a50ac6d0c55713f003a34af917ce71ad628f626

Mist 1.3 packaged and ready to go: https://github.com/ninxsoft/Mist/releases
